rust_src/src/multibyte.rs

```diff
@@ -88,13 +92,89 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions. This is a convenience function for
+    /// `self.display_width(None)`
     pub fn width(self) -> usize {
-        unsafe { lisp_string_width(self.into(), -1, ptr::null_mut(), ptr::null_mut()) as usize }
+        let (width, _) = self.display_width(None);
+        width
+    }
+
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions.
+    ///
+    /// With `precision` argument, return the width of longest substring that
+    /// doesn't exceed `precision`, and the number of characters and bytes it
+    /// contains in the returned tuple.
+    pub fn display_width(self, precision: Option<usize>) -> (usize, Option<(usize, usize)>) {
+        // Manually determine if string is unibyte (lets us ignore multibyte
+        // handling in more cases).
+        let len = self.len_chars() as usize;
+        let multibyte = self.len_chars() < self.len_bytes();
```

Comment:
Can we use a method here? That way if the definition of multibyte changes we are not hunting the code base for it?

Reply:
Maybe, but I'm not sure. The reason the `is_multibyte` method isn't used is so that in cases where strings are marked as multibyte but their contents are still unibyte, we can skip checking for unibyte characters. Maybe a `is_multibyte_content` method that does this check could be added (the actual name is up for bikeshedding of course).

Reply:
If the String is marked "multibyte" shouldn't the code trust that and proceed accordingly?